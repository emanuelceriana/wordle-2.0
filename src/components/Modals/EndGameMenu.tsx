import { useContext } from "react";
import { IAppContext, AppContext } from "../../context/AppContext";
import { Modal } from "./Modal";

export const EndGameMenu = () => {
  const { isEndGame, isWinner, activeRowIdx, randomWord, restartGame } =
    useContext<IAppContext>(AppContext);

  return (
    <Modal isOpen={isEndGame} showCloseButton={false}>
      <div data-testid="endGameMenu">
        {isWinner ? (
          <>
            <h3>You Win! ðŸ¥³ </h3>
            <p>{`You did it in just ${activeRowIdx} tries`}</p>
          </>
        ) : (
          <>
            <h3>You Lose! ðŸ˜°</h3>
            <p>
              The word was: <strong>{randomWord.toUpperCase()}</strong>
            </p>
          </>
        )}
        <div>
          <button onClick={restartGame}>
            {isWinner ? "Play Again?" : "Another Try?"}
          </button>
        </div>
      </div>
    </Modal>
  );
};
