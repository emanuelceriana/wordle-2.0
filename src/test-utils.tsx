import React, { ReactNode } from "react";
import { render } from "@testing-library/react";
import { AppContext } from "./context/AppContext";

const MockProvider: React.FC<{
  children: ReactNode | ReactNode[];
  value: object;
}> = ({ children, value }) => (
  <AppContext.Provider value={value}>{children}</AppContext.Provider>
);

export function renderWithContext(element: React.ReactElement, value: object) {
  return render(<MockProvider value={value}>{element}</MockProvider>);
}

export function withContext(element: React.ReactElement, value: object) {
  return <MockProvider value={value}>{element}</MockProvider>;
}
