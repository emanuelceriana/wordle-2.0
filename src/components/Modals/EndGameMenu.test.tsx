import { EndGameMenu } from "./EndGameMenu";
import { renderWithContext } from "../../test-utils";

interface MockValueContext {
  isEndGame: boolean;
  isWinner: boolean;
  activeRowIdx: number;
  randomWord: string;
  restartGame: () => void;
}

describe("<EndGameMenu />", () => {
  const restartGame = jest.fn();

  let mockValueContext: MockValueContext = {
    isEndGame: true,
    isWinner: true,
    activeRowIdx: 0,
    randomWord: "test",
    restartGame,
  };

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(<EndGameMenu />, mockValueContext);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be closed if is not end of the game", () => {
    mockValueContext.isEndGame = false;

    const { queryByTestId } = renderWithContext(
      <EndGameMenu />,
      mockValueContext
    );

    const endGameMenu = queryByTestId("endGameMenu");

    expect(endGameMenu).toBe(null);
  });
  it("should be opened if it is the end of the game", () => {
    mockValueContext.isEndGame = true;

    const { getByTestId } = renderWithContext(
      <EndGameMenu />,
      mockValueContext
    );

    const endGameMenu = getByTestId("endGameMenu");

    expect(endGameMenu).toBeDefined();
  });

  it("should show the winner message if the user wins", () => {
    mockValueContext.isEndGame = true;
    mockValueContext.isWinner = true;
    const winnerMessage = "You Win! ðŸ¥³";

    const { getByText } = renderWithContext(<EndGameMenu />, mockValueContext);

    const winnerMessageElement = getByText(winnerMessage);

    expect(winnerMessageElement).toBeInTheDocument();
  });

  it("should show the loser message if the user lose", () => {
    mockValueContext.isEndGame = true;
    mockValueContext.isWinner = false;
    const loserMessage = "You Lose! ðŸ˜°";

    const { getByText } = renderWithContext(<EndGameMenu />, mockValueContext);

    const loserMessageElement = getByText(loserMessage);

    expect(loserMessageElement).toBeInTheDocument();
  });
});
