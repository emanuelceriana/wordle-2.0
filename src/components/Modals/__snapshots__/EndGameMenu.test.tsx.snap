// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EndGameMenu /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="modalBg"
  />
  <div
    class="modalBg__modal"
  >
    <div
      class="modalBg__modal-title"
    />
    <div
      class="modalBg__modal-body"
    >
      <div
        data-testid="endGameMenu"
      >
        <h3>
          You Win! ðŸ¥³ 
        </h3>
        <p>
          You did it in just 0 tries
        </p>
        <div>
          <button>
            Play Again?
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
